pkg/controller/control.go

```diff
@@ -166,15 +169,28 @@ func (c *Controller) addReplica(address string, snapshotRequired bool, mode type
 	return c.addReplicaNoLock(newBackend, address, snapshotRequired, mode)
 }
 
+// Snapshot will try to freeze the filesystem of the volume if possible
+// and will fallback to a system level sync in all other cases
 func (c *Controller) Snapshot(name string, labels map[string]string) (string, error) {
-	// We perform a system level sync without the lock. Cannot block read/write
-	// Can be improved to only sync the filesystem on the block device later
+	log := logrus.WithFields(logrus.Fields{"volume": c.Name, "snapshot": name})
 	if ne, err := iutil.NewNamespaceExecutor(util.GetInitiatorNS()); err != nil {
-		logrus.Errorf("WARNING: continue to snapshot for %v, but cannot sync due to cannot get the namespace executor: %v", name, err)
+		log.WithError(err).Error("WARNING: cannot get namespace executor for volume sync/freeze")
 	} else {
-		if _, err := ne.ExecuteWithTimeout(syncTimeout, "sync", []string{}); err != nil {
-			// sync should never fail though, so it more like due to the nsenter
-			logrus.Errorf("WARNING: continue to snapshot for %v, but sync failed: %v", name, err)
+		// we always try to unfreeze the volume after finishing the snapshot process
+		// if the user manually froze the filesystem this will lead to an unfreeze but it's better than
+		// potentially having a left over frozen volume.
+		defer unfreezeVolume(ne, c.Name)
+		isFrozen, err := freezeVolume(ne, c.Name)
+		if err != nil {
+			log.WithError(err).Error("WARNING: failed to freeze the volume filesystem falling back to sync")
+		}
+
+		if !isFrozen {
```

Comment:
We may need to be careful about the error from `unfreezeFilesystem`. See the comment in `unfreezeFilesystem` 

```
syncRequired := false
mountPoint, err := findVolumeMountPoint()
if err != nil || mountPoint == "" {
    log()
    syncRequired := true
}
if mountPoint != "" { 
    defer func() {
        // make sure there is no error here if the below freezing failed
        err = unfreeze()
    }()
    if err := freeze();err != nil {
        log()
        syncRequired := true
    }
}
if syncRequired {
    ......
}
```

Reply:
What is your concern about looking for the mount point again?

Reply:
I mean, you will do mount point finding twice in `freezeVolume()` and `unfreezeVolume`, which is unnecessary.
`freeze` and `unfreeze` here means the operations to the filesystem, rather than invoking `freezeVolume()` and `unfreezeVolume`

Reply:
I find it better to search for the mount point twice, since if for whatever reason the mountpoint has changed (old one removed, new one added) we would still unfreeze the filesystem. While in the case you are proposing, the freeze would succeed but the unfreeze would fail. The cost of the extra findmnt should be small since it's only done as part of the snapshot operation.